Directory.Build.props

```diff
@@ -56,5 +56,27 @@
     <HumanizerVersion>2.14.1</HumanizerVersion>
     <MdocPackageVersion Condition=" '$(MdocPackageVersion)' == '' ">5.9.3</MdocPackageVersion>
   </PropertyGroup>
+  
+  <PropertyGroup>
+    <!-- 
+      Use $(TreatWarningsAsErrors) for 'Release' builds. We have grandfathered some projects that have existing warnings,
+      but we would like to go ahead and prevent any other projects from getting new warnings.
+    -->
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'assembly-store-reader.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'decompress-assemblies.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'jnienv-gen.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Microsoft.Android.Sdk.ILLink.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Microsoft.Android.Templates.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Mono.Android.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Mono.Android.NET-Tests.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'MSBuildDeviceIntegration.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'TestRunner.Core.NET.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Xamarin.Android.Build.Tasks.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Xamarin.Android.Build.Tests.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Xamarin.Android.JcwGen-Tests.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Xamarin.Android.NUnitLite.NET.csproj' ">true</_AllowProjectWarnings>
+    <_AllowProjectWarnings Condition=" '$(MSBuildProjectFile)' == 'Xamarin.ProjectTools.csproj' ">true</_AllowProjectWarnings>
+    <TreatWarningsAsErrors Condition=" '$(Configuration)' == 'Release' AND $(_AllowProjectWarnings) != 'true' ">true</TreatWarningsAsErrors>
```

Comment:
When I build locally, I'm not adding `-p:Configuration=Release` and I'd want to be able to catch these issues before pushing to CI and waiting for results. I wonder if we should enable this for both Debug and Release? Do we have more warnings that need fixing in Debug mode or is there some other reason to not include both default configs?

Reply:
I use Release builds locally also since allot of our unit tests only work in a Release build.

Reply:
The thought was that when you are writing early code you're just first trying to get something working, and having to fix NRT warnings or such takes you out of the flow.  So only enforcing warnings on `Release` was what we went with in JI:

https://github.com/dotnet/java-interop/blob/d5dfa0aac009ac4a8525f35a21dd469b8ccbc155/Directory.Build.props#L6

However, not seeing errors until CI can also be annoying.  Especially since `android` CI takes much longer than JI CI.

There are other options we could consider:
- Keying it off of `$(RunningOnCI)` instead of `Release`.
- Additionally adding a local opt-in property that could be enabled.  This could be set as an environment variable on ones local machine if someone prefers to always have warnings as errors enabled.

Reply:
Happy to be overruled here, it sounds like keeping this as currently implemented is preferable by most

Reply:
I'm in favor of doing something that would help us slowly solve the warnings over time.

@jpobst maybe you can use `$(RunningOnCI)=true` for these checks for now?

> I often use warnings to "debug" stuff, it would break workloads like that. For Debug, perhaps we should add a private property to enable warnings-as-errors?

@grendello you can set `$(TreatWarningsAsErrors)` for this. We still have `Configuration.Override.props` if you want to set it that way, or an env var should also work.

Reply:
> @grendello you can set $(TreatWarningsAsErrors) for this. We still have Configuration.Override.props if you want to set it that way, or an env var should also work.

I don't think this will work _today_, as it would also enable it for all the projects with existing warnings.  We would need a custom one for now to hit the grandfathering logic in this PR.  But ideally we would eventually be able to just use the global one.

Reply:
Updated to use `$(RunningOnCI)`.

Additionally, added `$(_AndroidTreatWarningsAsErrors)` to allow local opt-in to errors.